### ISSUE TITLE
Question about Stage-1 pretraining: paper says fast-token CE loss, but released training code seems to use AML / flow-matching loss

### ISSUE THREAD

**Issue by YifuXu1127** [amap-cvlab/ABot-Manipulation#10]
Hi ABot-M0 authors,

Thank you for open-sourcing ABot-M0. I have been reading both the paper and the released code, and I am confused about the implementation of **Stage-1 pretraining**.

From the paper, my understanding is that:

- In the two-stage training paradigm, **Stage 1** uses a **fast-token-based action classification / cross-entropy loss**
- Then **Stage 2** performs downstream SFT / task-specific fine-tuning

However, when I check the released training code, the current `Pretrain` pipeline seems to optimize an **AML / flow-matching style action loss** instead of an explicit fast-token CE loss.

Here is why I am confused:

1. The paper describes Stage 1 as fast-token CE pretraining
   - arXiv: https://arxiv.org/abs/2602.11236

2. The released pretraining script points to the normal training entry:
   - `examples/Pretrain/run_pretrain.sh`

3. In the training loop, the total loss seems to be only `action_loss`
   - `ABot/training/train.py`
   - specifically:
     ```python
     action_loss = output_dict["action_loss"]
     total_loss = action_loss
     ```

4. In `ABot/model/framework/ABot_M0.py`, `action_loss` is produced by the action model head:
   - `ABot/model/framework/ABot_M0.py`

5. In `ABot/model/modules/action_model/AML_ActionHeader.py`, this head appears to implement an AML / flow-matching objective on continuous actions

At the same time, I also noticed that the codebase still contains traces of the fast-token path:

- `ABot/model/modules/vlm/QWen3.py` has logic for:
  - action token ranges
  - `solutions`
  - constructing `labels` for token-level supervision

But in the released `Pretrain` training path, this `solutions -> labels -> CE loss` branch does not seem to be connected to the main training loop.

So I would like to ask:

1. Does the released `examples/Pretrain/run_pretrain.sh` reproduce the **paper Stage-1 pretraining** exactly?
2. If yes, where is the **fast-token CE loss** implemented in the released code?
3. If not, is the released code a later unified AML-based training version, while the paper describes an earlier / different Stage-1 pipeline?
4. Is any Stage-1 fast-token pretraining code missing from the current public release?

I am mainly asking for clarification so I can correctly understand and reproduce the training pipeline.

Thanks again for your work.

**Comment by YandanYang** [comment]
Hi, thanks for the careful reading — your understanding of the released code is correct.

**Short answers:**
1. `examples/Pretrain/run_pretrain.sh` does **not** reproduce the paper’s Stage-1 as a fast-token CE objective.
2. In the public release, the training objective is the **AML / continuous action** loss from `AML_ActionHeader` (via `ABot_M0.forward` → `action_loss`).
3. The fast-token helpers in `QWen3.py` (`solutions` → action-token labels) are **legacy / unused** in the current Pretrain path (`solutions` is never passed).
4. The paper’s Stage-1 wording refers to a **discretized FAST-token CE stage** used in our earlier / VLM-side pipeline; that full CE training path is **not wired into the current public Pretrain scripts**. The released Pretrain recipe is the **unified AML-based** training for VLM + action expert.

Hope this helps!
